sdk/src/utils/hash_utils.rs

```diff
@@ -174,7 +174,92 @@ pub fn hash_by_alg(alg: &str, data: &[u8], exclusions: Option<Vec<Exclusion>>) -
     }
 }
 
-// verify the hash using the specifiied alogrithm
+// Return hash bytes for assset using desired hashing algorithm.
+pub fn hash_asset_by_alg(
+    alg: &str,
+    asset_path: &Path,
+    exclusions: Option<Vec<Exclusion>>,
+) -> Result<Vec<u8>> {
+    use Hasher::*;
+    let mut hasher_enum = match alg {
+        "sha256" => SHA256(Sha256::new()),
+        "sha384" => SHA384(Sha384::new()),
+        "sha512" => SHA512(Sha512::new()),
+        _ => {
+            warn!(
+                "Unsupported hashing algorithm: {}, substituting sha256",
+                alg
+            );
+            SHA256(Sha256::new())
+        }
+    };
+
+    let mut data = File::open(asset_path)?;
+    let data_len = data.seek(SeekFrom::End(0))?;
+    data.seek(SeekFrom::Start(0))?;
+
+    let ranges = match exclusions {
+        Some(mut e) if !e.is_empty() => {
+            // hash data skipping excluded regions
+            // sort the exclusions
+            e.sort_by_key(|a| a.start());
+
+            // verify structure of blocks
+            let num_blocks = e.len();
+            let exclusion_end = e[num_blocks - 1].start() + e[num_blocks - 1].length();
+            let data_end = data_len - 1;
+
+            // if not enough range we will just calc to the end
+            if data_len < exclusion_end as u64 {
+                return Err(Error::BadParam(
+                    "The exclusion range exceed the data length".to_string(),
+                ));
+            }
+
+            //build final ranges
+            let mut ranges = RangeSet::<[RangeInclusive<u64>; 1]>::from(0..=data_end);
+            for exclusion in e {
+                let end = (exclusion.start() + exclusion.length() - 1) as u64;
+                let exclusion_start = exclusion.start() as u64;
+                ranges.remove_range(exclusion_start..=end);
+            }
+
+            ranges
+        }
+        _ => {
+            let data_end = data_len - 1;
+            RangeSet::<[RangeInclusive<u64>; 1]>::from(0..=data_end)
+        }
+    };
+
+    // hash the data for ranges
+    for r in ranges.into_smallvec() {
+        let start = r.start();
+        let end = r.end();
+        let mut chunk_left = end - start + 1;
+
+        // move to start of range
+        data.seek(SeekFrom::Start(*start))?;
+
+        loop {
+            let mut chunk = vec![0u8; std::cmp::min(chunk_left as usize, MAX_HASH_BUF)];
```

Comment:
If this becomes a perf bottleneck, we might be able to allocate this buffer once at start of range or start of function and then use slices within the buffer as the argument for `hasher_enum.update()`.